model/headers/headers.go

```diff
@@ -214,10 +215,15 @@ func (h Header) MarshalBinary() (data []byte, err error) {
 		return nil, err
 	}
 	for _, headerPacket := range h.HeaderPackets {
+		if h.Nonces != nil {
+			headerPacket.Nonce = h.Nonces[0]
+			h.Nonces = h.Nonces[1:]
+		}
 		marshalledHeaderPacket, err := headerPacket.MarshalBinary()
 		if err != nil {
 			return nil, err
 		}
+		h.Nonces = append(h.Nonces, headerPacket.Nonce)
```

Comment:
Maybe copy `h.Nonces` to a local variable and operate on that instead of removing and adding back?